Allow selecting nodes by CIDR policy

squeed opened the pull request:
This PR adds a new option, `policy-cidr-selects-nodes`, which means that cluster nodes can be selected by CIDR network policies. Normally nodes are only accessible via `remote-node` entity selectors.

**How it works:**
We already have the notion of an "identity scope" -- if the top 8 bits of the 32-bit identity are 0, we assume the ID is global. If the top bits are 0x01, then the ID is a local CIDR identity. This PR declares the scope 0x02 to be for node identities.

A separate identity scope is needed because the datapath needs to be able to quickly determine if a given identity corresponds to a remote node. Previously, all nodes had reserved identities 4 or 7. By using a separate identity scope, a fast bit comparison can be done without any additional map lookups.

So, most of this PR consists of refactoring commits to prepare for scoped local identity allocators. Then, a few commits at the end enable the behavior, switching nodes to the new identity scope and attaching `cidr:` labels to the local-node identity.

---

Fixes: https://github.com/cilium/cilium/issues/20550

squeed committed:
identity: Add Entity-CIDR Selectability

- identity: introduce concept of identity scopes

  We already use the high-8 bits of the numeric identity space to indicate
  what "kind" of identity is represented. This PR makes this more
  explicit, and defines it as an identity "scope".

  We will be adding another identity scope: remote-node, in future
  commits.

  No behavior changed, only identifiers.

- identity: prepare for scoped allocators

  Since we would like to support multiple local identity scopes, it means
  that the local identity allocator needs to be scope-aware.

  This change restructures the code a bit to make it clear that local
  allocators have a specific scope.

  This does not change any behavior.

- identity / bpf: reserve 0x02 for the remote-node scope

  This sets aside the identity scope of 0x02 for remote nodes. This scope
  is currently unused, but will be in future commits.

  It also updates the datapath to treat this as a remote node.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>
Signed-off-by: Nathan Sweet <nathanjsweet@pm.me>

squeed committed:
config: add policy-cidr-match-mode option

This option will be used in a future commit to enable selecting nodes
via CIDR selectors in Network Policy

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
identity: Add Node-CIDR Selectability

- identity: Move nodes to new identity scope (when enabled)

  Rather than assigning all nodes the remote-node identity, allocate them
  from the new remote-node local allocator. This functionality is only
  used when the option policy-cidr-match-world includes remote-node.

- ipcache / nodemanager: set / restore CIDR labels on nodes in ipcache

  We would like to be able to select nodes by CIDR
  selectors. So, we need to make sure the cidr labels are not forcibly
  removed when policy-cidr-match-mode=remote-node is set.

  Likewise, we need to supply the labels when inserting the node's prefix
  from the nodemanager. Additionally, we need to "backfill" the labels
  when restoring from an old ipcache.

  Now that remote nodes are allocated IDs from their own scope, the
  datapath can still route packets correctly, and it is safe to add a CIDR
  label and allocate an identity accordingly.

- test: tolerate partial reads when filtering by json

  It might be the case that we read a partial line when scanning for
  hubble flows. This JSON parse error causes the entire test to fail. We
  should just retry in that case.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>
Signed-off-by: Nathan Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
identity: Add Host-CIDR Selectability

- identity: Properly Assign Host Identity to IPs

  When the option "PolicyCIDRSelectsNodes" is set to true then
  we need to make sure to label all host IPs with their CIDR labels
  as well so that they are selectable.

  When nodemanager adds or updates the local host it is on, the
  IP addresses associated with it are added as labels to the host identity
  before it is initialized for the IPCache entry.

- policy: Unit Tests For Cluster Entity CIDR Selection

  When cluster entities are CIDR selectable via the
  PolicyCIDRMatchMode option, then test for these
  entities being selectable in a policy via a CIDR
  rule.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

squeed commented on bpf/lib/identity.h:
```diff
@@ -23,12 +23,18 @@ static __always_inline bool identity_is_remote_node(__u32 identity)
-	return identity == REMOTE_NODE_ID || identity == KUBE_APISERVER_NODE_ID;
+	return identity == REMOTE_NODE_ID ||
+		identity == KUBE_APISERVER_NODE_ID ||
+		(identity & 0xFF000000) == 0x02000000;
```
Should this be a #define somewhere? Probably.

joestringer replied:
For lack of a better place, top of this file would be fine. It is `identity.h` after all. Descriptive names that are the `CAPS_SNAKE_CASE` versions of the corresponding Go `CamelCase` variables is a common pattern we use.

squeed replied:
Done

squeed commented on pkg/ipcache/ipcache.go:
```diff
@@ -550,6 +550,36 @@ func (ipc *IPCache) RemoveIdentityOverride(cidr netip.Prefix, identityLabels lab
+//
+// Calling this function after startup is complete (and thus the prefix has almost certainly had an identity allocated
+// for it) will not be useful; it is only effective before startup when the ipcache is paused.
+func (ipc *IPCache) AllocateIdentity(ctx context.Context, prefix netip.Prefix, existing identity.NumericIdentity) (*identity.Identity, error) {
```
@joestringer let me know what you think about this API. The goal is to actually expose the API needed for restoration ("please reserve this cidr -> id mapping") rather than a sketchy mis-use of `AllocateCIDRs` and the reference-counting bugs we've had. If this works out, we can also transition the non-node CIDR restoration logic to use the same API.

squeed replied:
Note that this relies on the ipcache being stopped (because k8s caches are still syncing). If this is not desirable, we could add a bit of metadata that force-pauses updates for a given prefix. Not perfect, but potentially interesting.

joestringer replied:
I'll continue to mull this over, but I wonder whether longer term the ipcache restore logic should become part of ipcache so that it can (a) dump previous ipcache and understand IP<->NumericIdentity mappings, then (b) pull state from k8s/kvstore, then (c) internally resolve the exact desired labels for each IP + desired NumericIdentity, attempt the allocation of that identity or fall back to another NumericIdentity if necessary.

As a step to allow restore logic and *only* restore logic to inject a "desired NumericIdentity" for a prefix, this seems like a practical way to get it done.

squeed replied:
I agree that we should consider making restoration part of the ipcache; all of this is tricky since we have to back-fill the labels based on the numeric identity. That is definitely beyond the scope of this change, however.

squeed replied:
FYI, I added a unit test for this function, making sure it acts as we expect.

joestringer replied:
After a deeper look, and particularly noticing the synchronous identity allocation aspect of this, I'm less convinced that this is the right approach. I _think_ there may be a simpler solution, so I'll take a stab at trying to describe the pitfalls I see and a potential alternative in an overall review.

chancez reviewed (approved):
Hubble parser related changes LGTM.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -226,11 +241,105 @@ func New(c Configuration, ipCache IPCache, nodeMetrics *nodeMetrics, healthRepor
+
+	// Dump existing ip-identity pairs from the ipcache
+	existingPrefixes := map[netip.Prefix]identity.NumericIdentity{}
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
```
Dumping a large map another time during startup seems like unnecessary work. I see at the top of the function that there's an option config check to protect this and try to isolate this logic, but I think there are other ways to achieve this without adding all of this logic to node manager. I'm also not entirely sure why the node manager needs to deal with so many deep details of the datapath. Depending directly on the ipcachemap seems like a layering violation.

squeed replied:
This has been done, resolving this comment.

joestringer commented on pkg/node/manager/cell.go:
```diff
@@ -63,6 +63,8 @@ type NodeManager interface {
 	// by sending arping periodically.
 	StartNeighborRefresh(nh datapath.NodeNeighbors)
+
+	ReleaseRestoredIdentities()
```
This function in this interface seems out of place and leaks details across layers. We already have code to handle allocate and release of the restored identities. Adding a new very specific piece of logic to handle this in the NodeManager interface doesn't seem ideal. IMO it's better to leverage the existing path, and if there are bugs in that path then let's squash them.

christarazi replied:
Agreed. I can understand the motivation to prefer separate restoration logic that's specific to node CIDR identities and avoid touching the existing logic out of fear of dealing with refcounting bugs, but the proposed approach actually doesn't fundamentally change the fact that there _still_ could be a refcounting bug already existing in the code. 

A change that introduces the minimal diff to account for node CIDR identities means the chance of regression or a bug would be smaller than the current proposed changes, which involve a few layers.

squeed replied:
This interface has been removed; resolving.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -555,7 +555,7 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
-			if nid.HasLocalScope() {
+			if nid.HasLocalCIDRScope() {
 				d.restoredCIDRs = append(d.restoredCIDRs, k.Prefix())
 				oldNIDs = append(oldNIDs, nid)
```
I think that if we push the remote identity restore logic into this restore logic, we can drop the additional costly ipcache restore iteration, remove all of the new nodemanager logic, and remove the extra new identity allocation+release path (+ associated changes to the core agent wait times)

```suggestion
			if nid.HasLocalCIDRScope() || (nid.Scope() == identity.IdentityScopeRemoteNode && option.Config.PolicyCIDRSelectsNodes)
			{
				d.restoredCIDRs = append(d.restoredCIDRs, k.Prefix())
				oldNIDs = append(oldNIDs, nid)
```

I'll point out corresponding changes elsewhere in my next comment which could facilitate this in a relatively minimal way while only slightly tweaking the ipcache internal logic which already has to deal with identity details around specific remote node labels.

(Though we do need to think carefully about restore cases here for restoring various permutations of versions, such as from main today without the feature to this PR, from this PR to this PR restarting, or when toggling the new flag on/off)

joestringer replied:
^ I believe that this small change should automatically enqueue the release of the identities per the existing logic for identity restore & release.

This won't directly associate the remote-node labels with the identity, but we can infer that by masking the identity in the ipcache below. Either way, soon enough the k8s watcher in node manager will also associate that label with the CIDR as well.

squeed replied:
Sure, I can look in to this. I'm not as confident in all the details of the existing code path -- from my experience the older code is a lot more fragile, since it doesn't make the reference counting explicit -- but it should be possible.

squeed replied:
> we do need to think carefully about restore cases here for restoring various permutations of versions

So, I've given this some thought. The interesting case is transitioning from "enabled" to "disabled", with a policy that selects `reserved:remote-node`. (Transitioning from "disabled" to "enabled" doesn't have any restoration implications; just that some IPs change identity -- something we already handle gracefully).


Restoring to an older version *or* transitioning to the disabled flag currently is almost identical:
1: bpf ipcache map has nodeip `(N -> A)`; bpf policymap selects `(A, 6, 7)` because 6 and 7 are *always* selected)
2: The k8s listeners are synced. The ipcache updates the bpf map to point `(N -> 6)`
3: Endpoints are regenerated: bpf code is updated. In the downgrade case, we now only understand id 6 and 7 to be remote nodes
4: BPF policymap is updated. It now only selects `(6, 7)`

Unfortunately, there is no barrier between `ipcache.InjectLabels()` succeeding and endpoint regeneration, so we are at some risk for dropped traffic between 2 and 3. This *probably* affects other subsystems as well, FWIW.

There might be some other synchronization that ensures the ipcache has finished injecting before we regenerate endpoints, but I couldn't find it.

squeed replied:
Update: I'm silly. We make a new ipcache map on restart every time. This should certainly be safe to upgrade / downgrade / change flags at will.

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -427,22 +427,37 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
+		(!hasRemoteNode || (option.Config.PolicyCIDRSelectsNodes && hasRemoteNode)) &&
+		(!hasLocalHost || (option.Config.PolicyCIDRSelectsNodes && hasLocalHost)) {
+		cidrLabels := cidrlabels.GetCIDRLabels(prefix)
+		lbls.MergeLabels(cidrLabels)
```
If we allocate the node identities the standard way using the old API via `AllocateCIDRs()` which calls into this function (per the previous comment), I suspect we could drop the entire last two commits and just handle the remote node label restore logic right here.

```suggestion
		lbls.MergeLabels(cidrLabels)
                if option.Config.PolicyCIDRSelectsNodes &&
                  restoredIdentity.Scope() == identity.IdentityScopeRemoteNode &&
                  !hasRemoteNode && {
                       lbls.MergeLabels(labels.LabelRemoteNode)
                       hasRemoteNode = true
               }
```

squeed replied:
yes, but only for the restore case, not for the new-node case.

nathanjsweet replied:
@squeed @joestringer The updated logic does not include logic for restoring host CIDR labels, which the referenced commit(s) contained. Why was it removed?

joestringer commented on pkg/ipcache/metadata.go:
```diff
@@ -427,22 +427,37 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
 		n := lbls.Remove(labels.LabelWorld)
 		n = n.Remove(labels.LabelWorldIPv4)
 		n = n.Remove(labels.LabelWorldIPv6)
-		n = n.Remove(cidrlabels.GetCIDRLabels(prefix))
```
I think there's a problem here, specifically in the case where the new flag is disabled.

I believe that `info.ToLabels()` above could return the full set of CIDR labels iff there is a remote node and a CIDR policy that also specifies the same prefix. Previously the code would have removed the CIDR labels because that's how the current logic handles CIDRs for remote nodes - the remote nodes are not selectable by policy. I'm concerned that removing this line here would actually cause a regression where CIDR policy will now select remote nodes even if the flag is off.

I suspect this is why the logic was previously the other way around, with the labels being first removed... then only in specific scenarios the target labels are re-added. In a sense, the `info.ToLabels()` is a relatively unvetted set of requests from a range of resources to assign specific labels to the prefix & generate a corresponding Identity. This function is the function that's enforcing the policy posture that determines how the various remote-node, CIDR labels are mixed into the Identity.

squeed replied:
This function is confusing. Specifically, there is an unwritten assumption that the metadata layer / InjectLabels is only called for the class of entities that _might_ have CIDR labels. Then, we remove the CIDR labels only in the case of `host` and `remote-node`. Confusingly, we then re-add them unless we see the labels `host`, `remote-node`, `health`, or `ingress`.

So that means `health` or `ingress` _might_ have CIDR labels, but they might not. Huh.

So there are four cases here

- the prefix had CIDR labels, but we removed them
- the prefix had CIDR labels, and we kept them
- the prefix did not have CIDR labels, and we didn't add them
- the prefix did not have CIDR labels, but we added them


I see your point; we might unnecessarily re-add labels in the case that they are already in lbls. I'll re-think the logic with this perspective.

squeed replied:
OK, I've looked at `resolveIdentity()` a bit more closely and the diff is far more minimal now.

joestringer commented on pkg/identity/numericidentity.go:
```diff
@@ -583,9 +598,14 @@ func iterateReservedIdentityLabels(f func(_ NumericIdentity, _ labels.Labels)) {
-	return (id & LocalIdentityFlag) != 0
+// HasLocalCIDRScope returns true if the identity is in the Local CIDR scope
+func (id NumericIdentity) HasLocalCIDRScope() bool {
+	return id.Scope() == IdentityScopeLocalCIDR
```
nit: If we think this syntactic sugar is useful, may as well create corresponding helpers for `IdentityScopeLocalCIDR` as well.

joestringer replied:
(Seems like the code is not 100% consistent about whether to rely on these helper functions or just do a direct comparison. Would be nice to make it consistent, regardless of which approach is used)

squeed replied:
yeah, I added the equivalent function to make it more consistent.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -226,11 +241,105 @@ func New(c Configuration, ipCache IPCache, nodeMetrics *nodeMetrics, healthRepor
+		m.ipcache.UpsertMetadata(prefix, source.Restored, restoredResource, labels.LabelRemoteNode)
+
+		// Now that the ipcache knows about the prefix, try to allocate the existing ID number for it.
+		id, err := m.ipcache.AllocateIdentity(ctx, prefix, nid)
```
Another concern here if the argument above doesn't sell you, is there a race condition here? Can we guarantee that this start function completes before the ipcache starts?

squeed replied:
Indeed, that is something that works *now* but it might not later.

I'm definitely convinced that identity restoration should ultimately be moved in to the ipcache. For now, I'll work on getting restoration in to the existing pattern.

joestringer commented on pkg/identity/numericidentity.go:
```diff
@@ -23,9 +23,24 @@ const (
+
+	// IdentityScopeRemoteNode is the tag in the numeric identity that identifies
+	// an identity to be a remote in-cluster node.
+	IdentityScopeRemoteNode = NumericIdentity(2 << 24)
```
Should we define min/max remote node scoped identities? Do we need the full 16 million?

joestringer replied:
Related: does the allocator bound allocations within this range?

squeed replied:
Yeah, the allocated is bounded, so if we have 16 million node CIDRs allocation will fail.

From a resource perspective, we don't need the full 16 million. However, we do need a fast datapath "is-identity-remote-node" function. Right now that's just a single bitmask & compare.

I could shrink the space if you like and make it a simple range comparison, but I'm not super concerned about running out of ID scopes -- we have 254 more.

joestringer replied:
I was idly wondering about bounding the section smaller (only allocate the first 65K or so) while retaining the mask, but then it would mismatch with the datapath. So yeah from that perspective maybe it's better to just leave this as-is for now, and if we run out in some hypothetical future then we can always come up with some migration techniques to reclaim bitspace.

joestringer reviewed (changes requested):
I mulled this over with Chris a bit today, mostly focusing on the way that the ipcache changes work and the allocation/release of the identity. I'm not sure we need the top two commits if we just push a couple of extra lines into the daemon and ipcache. I also think that introducing another potentially-costly dump of ipcache in the node manager is a bit of a layer violation and introduces a fair bit of extra code without a direct need. I've suggested what I naively think seems like a possible alternative below, completely untested.

nebril commented on pkg/identity/cache/allocator.go:
```diff
@@ -275,9 +278,9 @@ func (m *CachingIdentityAllocator) Close() {
-		// to ensure we don't panic with 'send on closed channel'.
 		m.localCIDRIdentities.close()
+		m.localNodeIdentities.close()
+		close(m.events)
```
Is it possible for new events to be sent to this channel after closing? Seems like the comment you removed was concerned about that.

Won't this channel be garbage collected after we lose references to it?

squeed replied:
It used to be that `m.localIdentities.close()` would stop sending new events and close the channel. Now it just stops sending new events; I needed to "lift" channel closing out or else we would close it twice.

This is only ever used for clustermesh anyways, so I'm 99% certain none of this is actually *used*.

nebril commented on pkg/identity/cache/local.go:
```diff
@@ -194,14 +194,10 @@ func (l *localIdentityCache) GetIdentities() map[identity.NumericIdentity]*ident
-		close(l.events)
-		l.events = nil
-	}
+	l.events = nil
```
Why no channel close here? Seems like only `localIdentityCache` uses this channel to send.

squeed replied:
right; now there are two caches that send events to the same channel. So, we have to move the channel close "up a level" to prevent a double-close.

nebril commented on test/helpers/cmd.go:
```diff
@@ -485,6 +496,10 @@ func (res *CmdRes) WaitUntilMatchFilterLineTimeout(filter, expected string, time
 		if err != nil {
+			if errors.Is(err, &JSONParseError{}) {
+				// We might have read a partial line; continue
+				return false
```
Is there a way to emit a warning for that case? Not sure if it makes sense as probably nobody will look at passing test output :/.

squeed replied:
Looking at the uses of this function, we always wait for flows to appear, so it's safe to ignore parse errors.

I'd like to find a way to warn but I'm not sure the test library has that.

nebril commented on pkg/node/manager/manager.go:
```diff
@@ -385,6 +386,22 @@ func (m *manager) nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels
+				for _, address := range n.IPAddresses {
+					addr, ok := ip.AddrFromIP(address.IP)
+					if ok {
+						var b int = 128
```
Super nitpicky but - can we change `b` to `byteCount`? Definitely non-blocking.

nebril commented on test/runtime/net_policies.go:
Is `reserved:host` label being retired completely? If no, maybe we should retain at least some of these test cases to make sure that `reserved:host` still works?

squeed replied:
no, not the label, just the numerical identity. Good question.

nebril commented on daemon/cmd/daemon_main.go:
```diff
@@ -1801,12 +1802,13 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 			d.restoredCIDRs = nil
 		}
+		// Release restored node CIDR identities as well.
+		d.nodeDiscovery.Manager.ReleaseRestoredIdentities()
```
Shouldn't we have similar logic to above releasing of CIDRs? As in, check that there are any restored identities and then release them after grace period instead of always calling `Sleep`?

squeed replied:
We do; that's what this function does. (It is going to go away based on Joe's feedback, however).

nebril reviewed (approved):
Thanks for the PR! Left some comments inline, I didn't really look at the big-picture of the change as Joe and Chris already seem to have started this conversation.

I don't think that any of my comments are blocking, so I am approving the PR. Still, would appreciate getting answers to questions left in comments.

nathanjsweet commented on pkg/node/manager/manager.go:
```diff
@@ -385,6 +386,22 @@ func (m *manager) nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels
 	if m.conf.RemoteNodeIdentitiesEnabled() {
 		if n.IsLocal() {
 			nodeLabels = labels.NewFrom(labels.LabelHost)
+			if option.Config.EnableRemoteNodeIdentity {
```
This option is wrong.
```suggestion
			if option.Config.PolicyCIDRSelectsNodes {
```

asauber reviewed (approved):
Approving for changes to config, I haven't reviewed the identity logic.

squeed committed:
install: add policyCIDRMatchMode option

This adds the ability to enable this NetworkPolicy feature wherein nodes
are selectable by CIDR selectors. Ordinarily this is not allowed.

This commit just wires up the option.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

christarazi commented on pkg/node/manager/manager.go:
```diff
@@ -455,10 +472,17 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+		if option.Config.PolicyCIDRSelectsNodes {
+			lbls = labels.NewFrom(nodeLabels)
+			lbls.MergeLabels(cidrlabels.GetCIDRLabels(prefix))
+		}
```
Why is this necessary if `nodeIdentityLabels()` already encodes CIDR labels for each node's IP

joestringer replied:
I was wondering why it's necessary if `resolveIdentity()` in ipcache already detects "remote-node" label and injects the CIDR labels there.

squeed replied:
To minimize the diff to `resolveIdentity()`, I only changed it to _stop removing_ CIDR labels. However, we still need to add them somewhere. This seems like the most reasonable place -- the code that sets labels on a node.

joestringer replied:
I think the PR as-is is doing a pretty reasonable job on this front. I'm not the biggest fan of spreading the logic to handle this `option.Config.PolicyCIDRSelectsNodes` check across node manager, ipcache core (ie `resolveIdentity()`), and the identity allocator.. but I don't have a particularly concrete view on how we could better abstract this to handle the various cases.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -427,18 +427,28 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
-		// avoid allocating a CIDR identity for an entity within the cluster.
+
+	// If we are restoring an identity in the remote-node identity scope, we need to unconditionally add remote-node and cidr labels back
+	if restoredIdentity.Scope() == identity.IdentityScopeRemoteNode {
```
Couldn't we short-circuit this if we check for the feature flag so that we avoid adding CIDR labels here? I'm thinking in the flag=disabled case, the restored identity will be have the remote node scope, but the feature is off so we shouldn't be adding CIDR labels to begin win.

joestringer replied:
Related: What happens if there's a restored identity and the flag is off, then this function whittles down the labels to end up with `remote-node`. Will the identity allocator below just end up ignoring the request to restore this specific identity number and return the `reserved:remote-node` identity `6`?

squeed replied:
> Couldn't we short-circuit this if we check for the feature flag so that we avoid adding CIDR labels here?

Sure. easy.

> Related: What happens if there's a restored identity and the flag is off...?

Then we allocate the correct identity; `AllocateIdentity()` unconditionally calls `LookupReservedIdentityByLabels()` first, which does the right thing.

squeed replied:
Either way, this is moot; we wont try and restore this if the flag is off.

christarazi replied:
Could you add a code comment as to why it's moot to check the feature flag? It seems like something someone will stubble across and prevent them from having to understand the rest implementation to come to the same realization.

christarazi replied:
@nathanjsweet This is unresolved ^

nathanjsweet replied:
@christarazi [I added a comment](https://github.com/cilium/cilium/pull/27464/files#diff-bb1165de1a65911d59ac7f9a002aa9f1e25c30e073547adec11b174a174fdca6R435) several lines down where the CIDR is added.

christarazi replied:
I meant that the code comment should explain why it's not necessary to additionally check the feature flag, as the original comment on this line suggested.

christarazi commented on pkg/ipcache/metadata.go:
```diff
@@ -427,18 +427,28 @@ func (ipc *IPCache) resolveIdentity(ctx context.Context, prefix netip.Prefix, in
+		// So, remove if necessary
+		if !option.Config.PolicyCIDRSelectsNodes {
+			n = n.Remove(cidrlabels.GetCIDRLabels(prefix))
+		}
```
This doesn't seem necessary considering my previous comment in this file

joestringer replied:
Not quite, I think it's still possible for `info.ToLabels()` to return CIDR labels in the case where the new flag is disabled and a user specifies a CIDR policy. See https://github.com/cilium/cilium/pull/27464/files#r1318149295 for further discussion.

joestringer replied:
LGTM now.

christarazi commented on pkg/ipcache/metadata.go:
I would like to see unit tests for the changes here. It's very easy to introduce a regression.

squeed replied:
I've added some already; I'll add a few more with the flag on and off.

squeed replied:
(While the unit tests don't call `resolveIdentity` directly, it has good coverage through `TestInjectLabels()`)

squeed replied:
I've added more tests to `TestInjectLabels()`. At this point I think we cover all the cases:

flag off, cidr + kube-apiserver => node + kube-apiserver (identity changes from local scope to id 7
flag off, node => id 6

flag on, both prefixes transition to remote-node ID scope.
Ensure that all IDs are in remote-node scope. Ensure that all labels are as expected.

Remove remote-node label, ensure everything transitions to local cidr scope, ensure labels are as expected.

joestringer commented on pkg/defaults/defaults.go:
```diff
@@ -544,6 +544,9 @@ const (
 	EnableK8sNetworkPolicy = true
+
+	// PolicyCIDRSelectsNodes enables support for selecting nodes via CIDR selectors
+	PolicyCIDRSelectsNodes = true
```
I know this commit already says DNM (do not merge), but just highlighting that the default here should be false.

squeed replied:
Yup; in the commit that introduces this field, the default is false. I will drop the DNM commit approximately when this PR is ready to merge.

Just trying to get CI signal for now.

squeed replied:
I've removed this as default.

joestringer commented on pkg/identity/identity.go:
```diff
@@ -180,18 +181,33 @@ func (pair *IPIdentityPair) PrefixString() string {
 		switch label.Source {
 		case labels.LabelSourceCIDR, labels.LabelSourceReserved:
-			needsGlobal = false
+			scope = IdentityScopeLocalCIDR
```
Seems a bit odd to define `LabelSourceReserved` as requiring CIDR scope, since reserved identities aren't allocated in the local CIDR range. 

I'm not sure the implications of this, but at the very least we should probably add a comment here that reserved identities are just being treated like local CIDR scope because the semantics are closer to LocalCIDR than they are to ScopeGlobal.

jrajahalme replied:
It seems wrong that all reserved identities would get lumped into the CIDR scope? Maybe a separate non-CIDR local scope is needed for them?

squeed replied:
Right; none of the behavior changes, it's just naming. I don't know of any other "reserved" identities that are local besides CIDR identities, but I agree the naming is poor.

How about ScopeLocal and ScopeRemoteNode? That at least makes some sense. (Local is confusing, it means local identities, not local destinations, but comments should clear that up).

joestringer replied:
I think this is where the desired behaviours don't 100% line up with the bitpatterns used in the datapath, since reserved identities are be `0x000000xx` (scope: global) but they should be treated more like local identities (`0x01xxxxxx`). This function only checks for presence of the reserved label, and we could actually allocate `0x01xxxxxx` identities for these label sets particularly if they also include CIDR labels.

Put another way, there are:
1. Reserved identities in the range 1-255 with only `reserved:foo` labels
2. Local identities in the range `0x01xxxxxx` with CIDR labels and without `reserved:foo` labels
3. Local identities in the range `0x01xxxxxx` with CIDR labels and with `reserved:foo` labels (new in this PR)

The previous version of this function was answering the question "Does this identity need global identity allocation?" and the answer is the same for all three. I wonder about whether (1) and (3) belong to the same "scope" or whether perhaps (1) should be its own special scope. But in practical terms, I don't think that any code is making the wrong assumption based on the new version of this code introduced in this PR.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -555,7 +555,7 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 			v := value.(*ipcachemap.RemoteEndpointInfo)
 			nid := identity.NumericIdentity(v.SecurityIdentity)
-			if nid.HasLocalScope() {
+			if scope := nid.Scope(); scope == identity.IdentityScopeLocalCIDR || scope == identity.IdentityScopeRemoteNode {
```
If someone installs with this PR and hence populates the ipcache with identities in the new range, then turns the flag off, I guess that means that this function will still allocate those previous IdentityScopeRemoteNode identities on startup here, then later the agent will release them, and they'll be unused right?

Just wondering whether we need to restore the `IdentityScopeRemoteNode` identities if the flag is off?

squeed replied:
We will not actually allocate those identities (because the allocator will reject them, since the `ScopeForLabels()` will mismatch. However, it's safer to short-circuit here. I'll add that.

squeed replied:
Done.

joestringer commented on pkg/node/manager/manager.go:
I don't really follow why the node manager needs to care about the CIDR labels specifics here. IPCache is responsible for identity allocation in this case. Conceptually I think that the node manager should be able to just express to the ipcache "There are remote nodes at IPs X, Y, Z.". Maybe we can have a special Metadata type that generates these labels inside the ipcache, or maybe we just have the fixed config check logic similar to what is currently proposed inside ipcache `resolveLabels()`. Given that ipcache handles identity allocation, it can take responsibility to ensure that when theres a "remote node identity allocation" that it includes the appropriate labels depending on the value of the new flag.

squeed replied:
Something, somewhere, needs to add CIDR labels to those prefixes -- either the ipcache or its callers. Given that the NodeManager is already responsible for setting labels on node prefixes, it seems natural that it should, well, set all the desired labels.

This also leaves us open to a possible future change where we allocate a single identity per node, and even potentially make nodes selectable by labels.

There is precedent for this w.r.t. CIDR policy, which also allocates CIDR labels for prefixes that have CIDR selectors.

The NodeManager already has to have this logic, since it is the only place that can do so for the `reserved:node` identity. It needs to know *all* addresses, we *cannot* fix it up in `resolveIdentity()`.

Ultimately, I personally prefer for `resolveIdentity()` to do minimal policy invariant enforcement rather than general label business logic. Labels should come from `UpsertMetadata()`

joestringer replied:
I see yeah. I was thinking from the perspective of trying to avoid spreading this logic inside the core of ipcache as well as node manager. I also don't particularly like putting it into `resolveIdentity()`. Over time I think that `info.ToLabels()` could potentially handle this by introducing a special metadata types that handles the particular logic for node labels generation for this case. That would at least abstract it away from the core of ipcache, since `ToLabels()` is responsible for collecting all the metadata for the prefix and generating the set of labels. However, I buy that the NodeManager will want to be inserting requests to associate the node's labels into ipcache soon enough, and this is just another set of labels. :+1:

joestringer commented on test/helpers/cmd.go:
I take it that this is unrelated to the PR and should be submitted separately to just make Cilium's CI more robust?

squeed replied:
It's needed because node flow exports are now are long enough to trigger some buffer somewhere, making CI fail due to partial JSON reads.

joestringer replied:
Is that a problem that users will encounter in production?

squeed replied:
I don't believe so. This is happening because we're streaming hubble flows over ssh and naively assuming we always read complete lines. I assume actual users aren't going to do this. If so, their code is already broken.

squeed replied:
This particular PR uncovers this issue in our test code because the specific flows we're dumping are now decently larger, and thus longer than some line buffer somewhere in the SSH path. Thus, we get a partial read which is (eventually) filled in.

All the tests look for an expected flow, retrying until they find one in the SSH line buffer. So, this change is safe, since it shouldn't cause any tests so incorrectly start passing.

christarazi replied:
@nathanjsweet This was previously part of a separate commit. I think we should keep it separate in case we want to backport this specific commit at some point in the future.

joestringer commented on pkg/policy/distillery_test.go:
```diff
@@ -1453,6 +1458,21 @@ var (
+		},
+	}}).WithEndpointSelector(api.WildcardEndpointSelector)
+
+	mapKeyL3UnkownIngress = Key{identity.IdentityUnknown.Uint32(), 0, 0, trafficdirection.Ingress.Uint8()}
```
Oh yeah, and nit here: `Unknown` has three `n`s. Similar typo elsewhere.
```suggestion
	mapKeyL3UnknownIngress = Key{identity.IdentityUnknown.Uint32(), 0, 0, trafficdirection.Ingress.Uint8()}
```

squeed replied:
Fixed.

jrajahalme commented on pkg/identity/identity_test.go:
```diff
@@ -81,6 +81,16 @@ func (s *IdentityTestSuite) TestRequiresGlobalIdentity(c *C) {
+
+	c.Assert(ScopeForLabels(labels.NewLabelsFromModel([]string{"k8s:foo=bar"})), Equals, IdentityScopeGlobal)
+}
+
```
Please add a test case for a non-CIDR reserved identity here, other than the `remote-node` identity.

squeed replied:
Are there any? I didn't find any examples.

joestringer replied:
https://docs.cilium.io/en/stable/gettingstarted/terminology/#special-identities

squeed replied:
Right, the docblock is explicit about needing to be called after `LookupReservedIdentityByLabels`. I'll think if there are any other relevant cases.

squeed replied:
Added a good number of cases.

jrajahalme commented on pkg/identity/numericidentity.go:
```diff
@@ -26,6 +26,7 @@ const (
 	// Identities also have scopes, which is defined by the high 8 bits.
 	// 0x00 -- global and reserved identities
```
Note that datapath code seems to require scope 0 for reserved identities?

squeed replied:
It's implicit in the identity reservation; by definition id 7 is in global scope, since the high 8 bits are 0.

joestringer replied:
I have mild concerns with this, because reserved identities are not like global identities in a few key ways - they're not globally allocated and we generally avoid sending them on the wire. During upgrades if there are new reserved identities introduced, the older agents in the cluster will not know about them. I'm not sure if that has any bearing on this line of code but I don't think we should just categorize that upper 8 bits == 0 has the same properties. Those reserved identities with the prefix `0x000000xx` are special.

EDIT: To be clear, I'm not suggesting any specific changes to the PR, just raising that the abstraction is not entirely clear to me, so it would be nice to figure out how to clarify the above, either in this PR or in a subsequent change.

squeed replied:
right; this code doesn't change any behavior w.r.t reserved identities (with the obvious exception of remote-node). This is just documenting some already-existing expectations in the datapath.

If you come up with a better name for `IdentityScopeGlobal`, I'm happy to change.

squeed replied:
You're right; it's funny. Reserved identities are "global" in that all (up-to-date) have the same view of them; but they don't come from any centralized identity store. (Unless you consider quay.io centralized :-) ). That's why I called them out as such in the code comment.

christarazi replied:
I think the middle ground here is to keep the current changes + a comment describing the brief difference for reserved identities.

christarazi replied:
@nathanjsweet This is also unresolved ^

nathanjsweet replied:
@christarazi Why? I added details about the nature of reserved identities to the comment and specifically called out the version incompatibility problem.

jrajahalme commented on pkg/identity/cache/cache.go:
```diff
@@ -252,6 +260,13 @@ func (m *CachingIdentityAllocator) LookupIdentityByID(ctx context.Context, id id
 
+	switch id.Scope() {
+	case identity.IdentityScopeLocalCIDR:
+		return m.localCIDRIdentities.lookupByID(id)
```
This is duplicate code for the previous code block, remove the previous code block?

squeed replied:
Fixed.

jrajahalme commented on pkg/identity/identity_test.go:
```diff
@@ -51,10 +52,6 @@ func (s *IdentityTestSuite) TestReservedID(c *C) {
-	i = GetReservedID("kube-apiserver")
-	c.Assert(i, Equals, NumericIdentity(7))
-	c.Assert(i.String(), Equals, "kube-apiserver")
-
```
This deletion seems surprising. Is this PR removing the reserved identity for "kube-apiserver"? What if kube-apiserver is not a cluster node?

squeed replied:
When this behavior is enabled, `kube-apiserver` is just a label. The special identity is no longer needed.

I'll add this test logic back, behind the correct flag.

squeed replied:
Removing this was an oversight anyways; fixed.

jrajahalme commented on pkg/testutils/identity/allocator.go:
```diff
@@ -68,44 +73,38 @@ func (f *MockIdentityAllocator) AllocateIdentity(_ context.Context, lbls labels.
+		if !ok {
+			id = identity.NumericIdentity(f.nextIDs[scope]) | scope
 		}
+		f.nextIDs[scope]++
```
TIL increment operator works on a map like this :-)

jrajahalme reviewed (approved):
Some questions but overall looks good to me.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1138,6 +1138,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.MarkHidden(option.EnableK8sNetworkPolicy)
 	option.BindEnv(vp, option.EnableK8sNetworkPolicy)
 
+	flags.Bool(option.PolicyCIDRSelectsNodes, defaults.PolicyCIDRSelectsNodes, "enable selecting in-cluster nodes via NetworkPolicy CIDR selectors")
```
Idly wondering about the semantics here. Consider we might choose to change ToCIDR semantics in future again to allow it to match arbitrary IPs (not currently planned, but it's an ongoing point of friction). An alternative we could consider is:

- `--policy-cidr-match-mode=world` (current)
- `--policy-cidr-match-mode=world,remote-node` (this PR)
- `--policy-cidr-match-mode=world,remote-node,cluster` (not planned)

(Above is using the canonical "entity" names for how policy CIDR statement should match)

Or assume CIDRs select world and just have a single toggle rather than identity list:

- `--policy-cidr-match-mode=external` / `auto` / `default` / `none` (current)
- `--policy-cidr-match-mode=nodes` (this PR)
- `--policy-cidr-match-mode=cluster` (not planned)

Making this a bit more generic could potentially save us from having to either change the flag or introduce additional flags in future.

squeed replied:
Seems reasonable. Right now, `--policy-cidr-match-mode=remote-node` would not work, since I'm not about to go disabling the CIDR path. But I could imagine a future wherein that does work.

I'll add that flag and patch up the config logic.

squeed replied:
This is done.

joestringer commented on Documentation/cmdref/cilium-agent.md:
I take it you'll follow up on clear documentation for this as well?

squeed replied:
Extremely :-)

joestringer reviewed (approved):
The current approach in changing ipcache LGTM. I skipped over the event channel changes in the allocator internals, so if that hasn't undergone scrutiny then that'd be a good focus point for additional review.

I raised a question on the exact user-facing flag for this, but I don't have a strong opinion on it. See the thread below for some thoughts and feel free to chime in.

I'd like to make sure that the testing change this PR adds in `test/helpers` to handle some json encoding thing isn't hiding some real bug. It's not obvious to me why adding extra labels to a identities would trigger an issue like that. Thread ~below~ [above](https://github.com/cilium/cilium/pull/27464#discussion_r1319236594).

Other than that, before release we need crystal clear documentation on the new flag and its implications, and this would benefit from some additional e2e testing.

For your outstanding questions:

> Who should be responsible for adding CIDR labels to node prefixes: the NodeManager or resolveIdentity().

If possible I'd like to avoid spreading the new option check across node manager, ipcache _and_ identity allocator, but I do think that the PR as-is is functionally correct and is probably at a local minima for complexity without having to significantly refactor identity handling in the agent.

> Should we rename IdentityScopeLocalCIDR to IdentityScopeLocal (and other similar names)? (Probably yes, it can theoretically have non-CIDR identities).

Could you provide an example of a non-CIDR identity? I have a mild preference for Local Identity, since adding CIDR in there seems a bit restrictive and I don't think that the CIDR labels are critically important to the semantics of those identities, the biggest difference is that they're only significant to the local node and cannot be transmitted on the wire with any meaning.

Finally, as per the merging guidelines, please do make sure to either resolve all comments or file a tracking issue to follow up on any unresolved comments if action is necessary.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -385,6 +386,22 @@ func (m *manager) nodeIdentityLabels(n nodeTypes.Node) (nodeLabels labels.Labels
+						if addr.Is4() {
+							maskLength = 32
+						}
+						prefix, err := addr.Prefix(maskLength)
```
I think we can skip the hardcoding of bitlengths?
```suggestion
						prefix, err := addr.Prefix(addr.BitLen())
```

christarazi commented on pkg/identity/identity_test.go:
```diff
@@ -308,6 +343,7 @@ func TestLookupReservedIdentityByLabels(t *testing.T) {
 			assert.NotNil(t, id)
 			assert.Equal(t, tt.want.id, id.ID)
 			assert.Equal(t, tt.want.labels, id.Labels)
+
```
Nit: unneeded newline

christarazi reviewed (commented):
A couple more of small comments and it LGTM. I also have some questions around the commits:

* "config: add policy-cidr-match-mode option" seems like it can contain "install: add policyCIDRSelectsNodes option ". Is there a need for them to be separate?
* Ditto for "ipcache / nodemanager: set / restore CIDR labels on nodes in ipcache" and "policy: Unit Tests For Cluster Entity CIDR Selection "

christarazi reviewed (commented):
Please go through the remaining unresolved conversations and ensure the feedback is implemented. Also I think a few commits can be squashed together as they don't quite make sense being separate, which could break bisectability. See my previous review for those commits.

christarazi reviewed (approved):
LGTM now as there no outstanding concerns. Please go through the remaining unresolved conversations, as I still see 7 as reported by the GH UI.
